### PR TITLE
Azure (ARM): Fix GovCloud Support

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
This PR fixes the ability of the Azure ARM cloud module to communicate with Azure GovCloud endpoints. Some other issues with the cloud module were also fixed, along with general cleanup from linting, etc.

### What issues does this PR fix or reference?
N/A

### Previous Behavior
- Cloud module was unable to hit endpoints for anything other than the Public Cloud.
- Userdata was not able to be passed during instance creation... it silently failed.
- Some of the functions didn't work from the command line (`salt-cloud -f <function>`)

### New Behavior
- Broke out some functionality into a utils module which can also be used by some execution and state modules which are currently being worked.
- Cloud module can now hit endpoints for any Microsoft hosted endpoint or custom Azure Stack endpoint
- Custom scripts can now be passed via VM extensions. Added some documentation to the main docstring outlining the uses.
- Fixed functions to work from the command line in order to list various resources accessible to the cloud provider configuration.
- Fixed some logic to ensure the cloud module works as expected.
- Rolled in start/stop actions from PR 45119
- Linted the modules to ensure prettyness...

### Tests written?
No

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by gtmanfred — APPROVED** [review]

**Review by gtmanfred — CHANGES_REQUESTED** [review]
Can you change the name of the option for the public key to `ssh_key_file` so that it matches what is used in our other cloud providers.

Also if you could add a note about it to the getting started doc for azurearm

https://docs.saltstack.com/en/latest/topics/cloud/azurearm.html

Thanks,
Daniel

**Comment by gtmanfred** [comment]
Actually, sorry, i am wrong, you were right.  it can be ssh_publickeyfile. ssh_key_file should be the private key that is used to match that public key and log into the server.

**Comment by nicholasmhughes** [comment]
@gtmanfred I was too quick on the draw... Net effect is updated documentation. Is that correct now?

**Comment by gtmanfred** [comment]
It looks like you just need to add using the `ssh_key_file` as the identity file for the ssh command.

it should be passed to the `bootstrap()` function as [key_filename](https://github.com/saltstack/salt/blob/2017.7/salt/utils/cloud.py#L323)

Thanks,
Daniel

**Comment by nicholasmhughes** [comment]
@gtmanfred bootstrapping works as expected using the `ssh_keyfile` parameter, which is pulled in via `saltify.py`. I updated to documentation to reflect this... is that acceptable?

https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/saltify.py#L383

**Review by gtmanfred — APPROVED** [review]

**Comment by gtmanfred** [comment]
Yup, that works for me.

Thanks,
Daniel

**Comment by rallytime** [comment]
re-run lint

**Comment by rallytime** [comment]
re-run docs

**Comment by rallytime** [comment]
@techhat Can you review this?

**Comment by nicholasmhughes** [comment]
@rallytime @techhat Anything I can do to help with testing/acceptance?

**Comment by rallytime** [comment]
@nicholasmhughes We're still waiting on a review from @techhat. But in the mean time, this PR has a merge conflict in it. Can you fix that and rebase this?

**Comment by nicholasmhughes** [comment]
@rallytime I think the conflicts should be resolved. Please let me know if there are any other actions.

**Comment by drawsmcgraw** [comment]
@rallytime @techhat anything more we can do to assist with the testing/merging?I can buy @techhat a coffee if it helps :)

File: doc/topics/cloud/azurearm.rst
Comment (techhat) [inline doc/topics/cloud/azurearm.rst]: `ssh_password` should not be removed from the example, since it is still supported. How about another example block for `ssh_publickeyfile`?

**Review by techhat — APPROVED** [review]
This is a really great set of changes! Thanks @nicholasmhughes!

**Comment by nicholasmhughes** [comment]
@techhat is the documentation change in line with what you were thinking?

**Comment by techhat** [comment]
@nicholasmhughes perfect, thanks!

**Pull request merged**
